extensions/warp-ipfs/src/store/document.rs

```diff
@@ -279,3 +284,93 @@ impl RootDocument {
         Ok(root_document)
     }
 }
+
+#[derive(Clone, Deserialize, Serialize)]
+pub struct FileAttachmentDocument {
+    pub name: String,
+    pub size: usize,
+    pub thumbnail: Cid,
+    pub file_type: FileType,
+    pub data: Cid,
+}
+
+impl FileAttachmentDocument {
+    pub async fn resolve_to_file(&self, ipfs: &Ipfs) -> Result<File, Error> {
+        let file = File::new(&self.name);
+        file.set_size(self.size);
+        file.set_file_type(self.file_type.clone());
+
+        let image: ImageDag = ipfs
+            .get_dag(self.thumbnail)
+            .timeout(Duration::from_secs(10))
+            .deserialized()
+            .await?;
+
+        file.set_thumbnail_format(image.mime.into());
+
+        let data = ipfs
+            .unixfs()
+            .cat(
+                self.thumbnail,
+                None,
+                &[],
+                false,
+                Some(Duration::from_secs(10)),
+            )
+            .await
+            .unwrap_or_default();
+
+        file.set_thumbnail(&data);
+
+        Ok(file)
+    }
+
+    pub fn download<'a, P: AsRef<Path>>(
+        &'a self,
+        ipfs: &'a Ipfs,
+        path: P,
+        members: &'a [PeerId],
+        timeout: Option<Duration>,
+    ) -> BoxStream<'a, Progression> {
```

Comment:
I think you don't need explicit lifetimes here:

```suggestion
    pub fn download<P: AsRef<Path>>(
        &self,
        ipfs: &Ipfs,
        path: P,
        members: &[PeerId],
        timeout: Option<Duration>,
    ) -> BoxStream<'_, Progression> {
```

Reply:
Tried it before but compiler calls for lifetime unless im overlooking something here

Reply:
Since this is commented out for now, this could probably be looked into at a later PR when the changes are implemented since I may need to make changes upstream too.

Reply:
> Tried it before but compiler calls for lifetime unless im overlooking something here

What's the error?